Add unit tests for functions within `.~/data/utils.js`

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Add unit tests for functions within `.~/data/utils.js`

- Add test cases for the returning `ids` field of `getReportQuery` function
- Add test cases for `getReportKey`, `calculateDelta`, and `mapReportFieldsToPerformance` functions

### Detailed test instructions:

1. Run `npm run test-unit` and make sure all these tests pass

### Changelog entry

eason9487 committed:
Add test cases for the returning  field of getReportQuery function

eason9487 committed:
Add test cases for getReportKey function

eason9487 committed:
Add test cases for calculateDelta function

eason9487 committed:
Add test cases for mapReportFieldsToPerformance function

eason9487 committed:
Rephrase descriptions for test cases within .~/data/utils.js

tomalec commented on js/src/data/utils.test.js:
```diff
@@ -65,4 +73,180 @@ describe( 'getReportQuery', () => {
+		forAnyCategoryAndDateReference( ( category, dateReference ) => {
+			const query = {};
+
+			expect(
+				getReportQuery( category, 'free', query, dateReference ).ids
+			).toBeUndefined();
```
:nail_care: If we are specing "should have no `ids` field", I'd rather use `not.toHaveProperty`
```suggestion
			expect(
				getReportQuery( category, 'free', query, dateReference )
			).not.toHaveProperty( 'id' );
```

tomalec replied:
Same applies for the ones below, `like `.toHaveProperty( 'ids', query.programs )` - it gives more meaningful errors when fail

eason9487 replied:
> If we are specing "should have no `ids` field", I'd rather use `not.toHaveProperty`

Agreed, that's the correct way to verify this case.

> it gives more meaningful errors when fail

Thanks! Updated by https://github.com/woocommerce/google-listings-and-ads/pull/856/commits/9e3533f5a3b3c9f82e7ba7e67a9ad9ac1f7fa86a.

tomalec commented on js/src/data/utils.test.js:
```diff
@@ -65,4 +73,180 @@ describe( 'getReportQuery', () => {
+		).toBe( query.programs );
+	} );
+
+	it( 'When the products report query contains `products`, should have transformed `ids` field', () => {
```
:nail_care: I'd explain what "transformation" is expected.
```suggestion
	it( 'When the products report query contains `products`, should have `ids` field, with values prepended with `gla_`', () => {
```

tomalec commented:
:scroll: :nail_care: If we aim for fuller coverage, the default values of `mapReportFieldsToPerformance` are not covered.

```
npm run test-unit -- js/src/data/utils.test.js --coverage

-----------|---------|----------|---------|---------|-------------------
File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
-----------|---------|----------|---------|---------|-------------------
All files  |     100 |    93.75 |     100 |     100 |                   
 data      |     100 |    93.55 |     100 |     100 |                   
  utils.js |     100 |    93.55 |     100 |     100 | 111-112           
 utils     |     100 |      100 |     100 |     100 |                   
  round.js |     100 |      100 |     100 |     100 |                   
-----------|---------|----------|---------|---------|-------------------

```

eason9487 committed:
Tweak a test description for getReportQuery

Address https://github.com/woocommerce/google-listings-and-ads/pull/856/files#r666472854

Co-authored-by: Tomek Wytrębowicz <tomalecpub@gmail.com>

eason9487 committed:
Use `toHaveProperty` to verify object property

Address https://github.com/woocommerce/google-listings-and-ads/pull/856#discussion_r666470995

eason9487 committed:
Tweak a test case description

eason9487 committed:
Add a test case for mapReportFieldsToPerformance to cover the default values

Address https://github.com/woocommerce/google-listings-and-ads/pull/856#issuecomment-876704793

eason9487 commented:
> If we aim for fuller coverage, the default values of `mapReportFieldsToPerformance` are not covered.

Added by https://github.com/woocommerce/google-listings-and-ads/pull/856/commits/836796374d9684ff351cde6192017605ff9ad1ea

tomalec reviewed (approved):
LGTM, thanks :)
